### PR TITLE
action_recognition_demo.py: use AsyncInferQueue

### PR DESCRIPTION
**Pull request by anzhella-pankratova**
The PR fixes two separate issues with action_recognition_demo:
- The usage of AsyncInferQueue instead of list of inference requests prevents the demo hanging for both CPU and GPU devices
- Adds the cancelation of all encoder and decoder infer requests before the demo ends. It helps to avoid accidental crashes on CI, when the demo has already finished inference

### CODE REVIEW

File: demos/action_recognition_demo/python/action_recognition_demo/models.py [hunk 1]
```diff
@@ -93,22 +95,27 @@ def __init__(self, model_path, core, target_device, num_requests, model_type):
...
+
     def async_infer(self, frame, req_id):
         input_data = {self.input_name: frame}
-        self.infer_requests[req_id].start_async(inputs=input_data)
+        self.infer_queue.start_async(input_data, req_id)
 
     def wait_request(self, req_id):
-        self.infer_requests[req_id].wait()
-        return self.infer_requests[req_id].results[self.output_tensor]
+        self.infer_queue.wait_all()
```
Comment (Wovchena) [inline demos/action_recognition_demo/python/action_recognition_demo/models.py:117]: Is there a way to wait only `req_id`?

Comment (anzhella-pankratova) [reply demos/action_recognition_demo/python/action_recognition_demo/models.py]: Yeah, we can get the infer request by id and then call wait() for it
Also, I updated gesture_recognition_demo with same changes

**Comment by anzhella-pankratova** [comment]
Jenkins please retry a build

**Comment by anzhella-pankratova** [comment]
Jenkins please retry a build

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py [hunk 2]
```diff
@@ -101,6 +101,9 @@ def __init__(self, encoder):
         self.encoder = encoder
         self.async_model = AsyncWrapper(self.encoder, self.encoder.num_requests)
 
+    def __del__(self):
+        self.encoder.infer_queue[0].cancel()
```
Comment (anzhella-pankratova) [inline demos/action_recognition_demo/python/action_recognition_demo/steps.py:105]: Should we call `cancel()` for all infer requests in infer queue or is it enough only for the first infer request? @Wovchena @VoronovaIntern

Comment (Wovchena) [reply demos/action_recognition_demo/python/action_recognition_demo/steps.py:105]: For all

**Comment by VoronovaIntern** [comment]
cancels all encoder and decoder requests before the demo ends, to avoid accidental crashes

**Comment by anzhella-pankratova** [comment]
Jenkins please retry a build

File: demos/action_recognition_demo/python/action_recognition_demo/pipeline.py [hunk 3]
```diff
@@ -57,7 +57,8 @@ def start(self):
 
     def join(self):
         self.input_queue.put(Signal.STOP)
-        self._thread.join()
+        if self._thread is not None:
```
[the remark this answers is not in the document]
Comment (VoronovaIntern) [reply demos/action_recognition_demo/python/action_recognition_demo/pipeline.py:60]: @Wovchena , there were moments when the demo fell on this

[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/action_recognition_demo/python/action_recognition_demo/pipeline.py:60]: Is there an explanation, why it failed?

**Comment by VoronovaIntern** [comment]
Jenkins please retry a build

**Comment by Wovchena** [comment]
Jenkins please retry a build

**Pull request merged**